sql/core/src/main/scala/org/apache/spark/sql/Observation.scala

```diff
@@ -0,0 +1,147 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import java.util.UUID
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.locks.{Condition, Lock, ReentrantLock}
+
+import org.apache.spark.sql.execution.QueryExecution
+import org.apache.spark.sql.util.QueryExecutionListener
+
+/**
+ * Not thread-safe.
+ * @param name
+ * @param sparkSession
+ */
+class Observation(name: String) {
+
+  private val lock: Lock = new ReentrantLock()
+  private val completed: Condition = lock.newCondition()
+  private val listener: ObservationListener = ObservationListener(this)
+
+  private var sparkSession: Option[SparkSession] = None
+
+  @transient private var row: Option[Row] = None
+
+  /**
+   * Attach this observation to the given Dataset.
+   * Remember to call `close()` when the observation is done.
+   *
+   * @param ds dataset
+   * @tparam T dataset type
+   * @return observed dataset
+   */
+  def on[T](ds: Dataset[T])(expr: Column, exprs: Column*): Dataset[T] = {
+    if (ds.isStreaming) {
+      throw new IllegalArgumentException("Observation does not support streaming Datasets")
+    }
+    register(ds.sparkSession)
+    ds.observe(name, expr, exprs: _*)
+  }
+
+  /**
+   * Wait for the first action on the observed dataset to complete and returns true.
+   * This method times out after the given amount of time and returns false.
+   *
+   * @param time timeout
+   * @param unit timeout time unit
+   * @return true if action complete within timeout, false on timeout
+   */
+  def waitCompleted(time: Long, unit: TimeUnit): Boolean = waitCompleted(Some(time), unit)
+
+  /**
+   * Get the observation results. This waits until the observed dataset finishes its first action.
+   * If you want to wait for the result and provide a timeout, use waitCompleted.
+   * Only the result of the first action is available. Subsequent actions do not modify the result.
+   */
+  def get: Row = {
+    waitCompleted(None, TimeUnit.SECONDS)
+    row.get
+  }
+
+  private def waitCompleted(time: Option[Long], unit: TimeUnit): Boolean = {
```

Comment:
shall we simply use `synchronized instead of lock/unlock?

Reply:
do we need to block? `completed.await` should be thread-safe?

Reply:
Right, I can `wait` and `notifyAll` on `this` when I use a `synchronized` block. Will replace the lock and condition with `synchronized`.

Not sure what you mean with `to block`? Calling into `await` / `wait`?